src/toxcore/dht_new/packet.rs

```diff
@@ -0,0 +1,938 @@
+            IpAddr::V6(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+// TODO: move it somewhere else
+impl FromBytes for Ipv4Addr {
+    named!(from_bytes<Ipv4Addr>, map!(count!(le_u8, 4), 
+        |v| Ipv4Addr::new(v[0], v[1], v[2], v[3])
+    ));
+}
+
+impl ToBytes for Ipv4Addr {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        let o = self.octets();
+        do_gen!(buf,
+            gen_be_u8!(o[0]) >>
+            gen_be_u8!(o[1]) >>
+            gen_be_u8!(o[2]) >>
+            gen_be_u8!(o[3]) 
+        )
+    }
+}
+
+// TODO: move it somewhere else
+impl FromBytes for Ipv6Addr {
+    named!(from_bytes<Ipv6Addr>, map!(count!(le_u16, 8), 
+        |v| Ipv6Addr::new(v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7])
+    ));
+}
+
+impl ToBytes for Ipv6Addr {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        let s = self.segments();
+        do_gen!(buf,
+            gen_le_u16!(s[0]) >>
+            gen_le_u16!(s[1]) >>
+            gen_le_u16!(s[2]) >>
+            gen_le_u16!(s[3]) >>
+            gen_le_u16!(s[4]) >>
+            gen_le_u16!(s[5]) >>
+            gen_le_u16!(s[6]) >>
+            gen_le_u16!(s[7]) 
+        )
+    }
+}
+
+/** `PackedNode` format is a way to store the node info in a small yet easy to
+parse format.
+
+It is used in many places in Tox, e.g. in `DHT Send nodes`.
+
+To store more than one node, simply append another on to the previous one:
+
+`[packed node 1][packed node 2][...]`
+
+Serialized Packed node:
+
+Length | Content
+------ | -------
+1      | [`IpType`](./.enum.IpType.html)
+4 or 16| IPv4 or IPv6 address
+2      | port
+32     | node ID
+
+Size of serialized `PackedNode` is 39 bytes with IPv4 node info, or 51 with
+IPv6 node info.
+
+DHT module *should* use only UDP variants of `IpType`, given that DHT runs
+solely on the UDP.
+*/
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct PackedNode {
+    /// Socket addr of node.
+    saddr: SocketAddr,
+    /// Public Key of the node.
+    pub pk: PublicKey,
+}
+
+/// Size in bytes of serialized [`PackedNode`](./struct.PackedNode.html) with
+/// IPv4.
+pub const PACKED_NODE_IPV4_SIZE: usize = PUBLICKEYBYTES + 7;
+/// Size in bytes of serialized [`PackedNode`](./struct.PackedNode.html) with
+/// IPv6.
+pub const PACKED_NODE_IPV6_SIZE: usize = PUBLICKEYBYTES + 19;
+
+/** Serialize `PackedNode` into bytes.
+
+Can be either [`PACKED_NODE_IPV4_SIZE`]
+(./constant.PACKED_NODE_IPV4_SIZE.html) or [`PACKED_NODE_IPV6_SIZE`]
+(./constant.PACKED_NODE_IPV6_SIZE.html) bytes long, depending on whether
+IPv4 or IPv6 is being used.
+*/
+impl ToBytes for PackedNode {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        do_gen!(buf,
+            gen_if_else!(self.saddr.is_ipv4(), gen_be_u8!(IpType::U4 as u8), gen_be_u8!(IpType::U6 as u8)) >>
```

Comment:
Nice! But also replace IpType::U4 into inline integer values. And remove struct IpType since we don't need it any more.